### PR TITLE
Add ssm patch support

### PR DESCRIPTION
**Pull request by jamengual**
## what
* move ssm to its own file
* fix invalid resource reference

## why
* patch fix

### CODE REVIEW

File: ssm.tf [hunk 1]
```diff
@@ -0,0 +1,45 @@
+module "label_ssm_patch_s3_log_policy" {
+  source  = "cloudposse/label/null"
+  version = "0.24.1"
+
+  attributes = ["ssm-patch-s3-logs"]
+  context    = module.this.context
```
Comment (aknysh) [inline ssm.tf:6]: enable = var.ssm_patch_enabled

File: main.tf [hunk 2]
```diff
@@ -1,5 +1,6 @@
 locals {
-  instance_count = module.this.enabled ? 1 : 0
+  enabled        = module.this.enabled
+  instance_count = local.enabled ? 1 : 0
   volume_count   = var.ebs_volume_count > 0 && local.instance_count > 0 ? var.ebs_volume_count : 0
```
Comment (aknysh) [inline main.tf:4]:
```suggestion
  volume_count   = local.enabled && var.ebs_volume_count > 0 ? var.ebs_volume_count : 0
```

File: main.tf [hunk 3]
```diff
@@ -1,5 +1,6 @@
 locals {
-  instance_count = module.this.enabled ? 1 : 0
+  enabled        = module.this.enabled
+  instance_count = local.enabled ? 1 : 0
```
Comment (aknysh) [inline main.tf:3]:
```suggestion
  
```

File: ssm_patch.tf [hunk 4]
```diff
@@ -0,0 +1,52 @@
+
+locals {
+  ssm_patch_log_bucket_enabled = local.ssm_enabled && var.ssm_patch_manager_s3_log_bucket != "" && var.ssm_patch_manager_s3_log_bucket != null
+  ssm_policy                   = var.ssm_patch_manager_iam_policy == null || var.ssm_patch_manager_iam_policy == "" ? "arn:aws:iam::aws:policy/AmazonSSMManagedInstanceCore" : var.ssm_patch_manager_iam_policy
+  ssm_enabled                  = local.enabled && var.ssm_patch_manager_enabled ? true : false
```
Comment (aknysh) [inline ssm_patch.tf:5]: move this to the top of `locals`

Comment (aknysh) [inline ssm_patch.tf:5]:
```suggestion
  ssm_enabled                  = local.enabled && var.ssm_patch_manager_enabled
```

File: ssm_patch.tf [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+}
+
+module "label_ssm_patch_s3_log_policy" {
+  source  = "cloudposse/label/null"
+  version = "0.24.1"
+
+  attributes = ["ssm-patch-s3-logs"]
```
Comment (aknysh) [inline ssm_patch.tf:12]: enabled = local.ssm_enabled

File: variables.tf [hunk 6]
```diff
@@ -319,3 +319,21 @@ variable "volume_tags_enabled" {
   default     = true
   description = "Whether or not to copy instance tags to root and EBS volumes"
 }
+
+variable "ssm_patch_manager_enabled" {
+  type        = bool
+  default     = false
+  description = "Whether to enable SSM Patch manager"
```
Comment (aknysh) [inline variables.tf:326]:
```suggestion
  description = "Whether to enable SSM Patch Manager"
```

File: variables.tf [hunk 7]
```diff
@@ -319,3 +319,21 @@ variable "volume_tags_enabled" {
...
+}
+
+variable "ssm_patch_manager_iam_policy" {
+  type        = string
+  default     = null
+  description = "IAM policy to allow Patch manager to manage the instance"
```
Comment (aknysh) [inline variables.tf:332]:
```suggestion
  description = "IAM policy to allow SSM Patch Manager to manage the instance"
```

**Review by aknysh — COMMENTED** [review]
looks good @jamengual , just a few nitpicks

**Comment by aknysh** [comment]
@jamengual 

I'd rename 

```
variable "ssm_patch_manager_iam_policy" {
  type        = string
  default     = null
  description = "IAM policy to allow Patch manager to manage the instance"
}
```

to 

```
variable "ssm_patch_manager_iam_policy_arn" {
  type        = string
  default     = null
  description = "IAM policy ARN to allow Patch Manager to manage the instance. If not provided, `arn:aws:iam::aws:policy/AmazonSSMManagedInstanceCore` will be used"
}
```

File: ssm_patch.tf [hunk 8]
```diff
@@ -0,0 +1,53 @@
+
+locals {
+  ssm_patch_log_bucket_enabled = local.ssm_enabled && var.ssm_patch_manager_s3_log_bucket != "" && var.ssm_patch_manager_s3_log_bucket != null
+  ssm_policy                   = var.ssm_patch_manager_iam_policy == null || var.ssm_patch_manager_iam_policy == "" ? "arn:aws:iam::aws:policy/AmazonSSMManagedInstanceCore" : var.ssm_patch_manager_iam_policy
```
Comment (aknysh) [inline ssm_patch.tf:4]:
```suggestion
  ssm_policy_arn                   = var.ssm_patch_manager_iam_policy_arn == null || var.ssm_patch_manager_iam_policy_arn == "" ? "arn:aws:iam::aws:policy/AmazonSSMManagedInstanceCore" : var.ssm_patch_manager_iam_policy_arn
```

File: ssm_patch.tf [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+  ssm_enabled                  = local.enabled && var.ssm_patch_manager_enabled
+}
+
+module "label_ssm_patch_s3_log_policy" {
+  enabled = local.ssm_patch_log_bucket_enabled
```
Comment (aknysh) [inline ssm_patch.tf:9]: source and version should be before any other attributes

**Review by aknysh — APPROVED** [review]
please see comments

**Comment by jamengual** [comment]
> please see comments

Fixed

File: main.tf [hunk 10]
```diff
@@ -1,7 +1,7 @@
 locals {
   enabled        = module.this.enabled
   instance_count = local.enabled ? 1 : 0
-  volume_count   = var.ebs_volume_count > 0 && local.instance_count > 0 ? var.ebs_volume_count : 0
+  volume_count   = local.enabled && var.ebs_volume_count > 0 ? var.ebs_volume_count : 0
```
Comment (aknysh) [inline main.tf:4]: why removing `local.instance_coun > 0` ?
Volumes are attached to instances, so if `instance_count = 0` we prob should not try to attach

**Review by aknysh — DISMISSED** [review]

**Review by aknysh — APPROVED** [review]

**Pull request merged**